pkg/auth/handler/saml/login_finish.go

```diff
@@ -20,9 +26,71 @@ func NewLoginFinishHandlerLogger(lf *log.Factory) *LoginFinishHandlerLogger {
 }
 
 type LoginFinishHandler struct {
-	Logger *LoginFinishHandlerLogger
+	Logger                     *LoginFinishHandlerLogger
+	SAMLService                HandlerSAMLService
+	SAMLSessionService         SAMLSessionService
+	AuthenticationInfoResolver SAMLAuthenticationInfoResolver
+	AuthenticationInfoService  SAMLAuthenticationInfoService
 }
 
 func (h *LoginFinishHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	authInfoID, ok := h.AuthenticationInfoResolver.GetAuthenticationInfoID(r)
+	if !ok {
+		h.Logger.Warningln("authentication info id is missing")
+		http.NotFound(rw, r)
+		return
+	}
 
+	authInfo, err := h.AuthenticationInfoService.Get(authInfoID)
+	if err != nil {
+		// It is unexpected that we've set the code in query, but turns out it does not exist.
+		// We have no idea how to return the error response to user yet,
+		// so we can only panic.
+		panic(err)
+	}
+
+	samlSession, err := h.SAMLSessionService.Get(authInfo.SAMLSessionID)
+	if err != nil {
+		// It is unexpected that we've set the session id in auth info, but turns out it does not exist.
+		// We have no idea how to return the error response to user yet,
+		// so we can only panic.
+		panic(err)
+	}
+
+	result := h.handleLoginResult(authInfo, samlSession)
+	result.WriteResponse(rw, r)
+}
+
+func (h *LoginFinishHandler) handleLoginResult(
+	authInfo *authenticationinfo.Entry,
+	samlSession *samlsession.SAMLSession,
+) (result httputil.Result) {
+	defer func() {
+		if e := recover(); e != nil {
+			e := panicutil.MakeError(e)
+			h.Logger.WithError(e).Error("panic")
+			// TODO(saml): Return a error
+			panic(e)
+		}
+	}()
+
+	authnRequest := samlSession.Entry.AuthnRequest()
+	authenticatedUserID := authInfo.T.UserID
+
+	resp, err := h.SAMLService.IssueSuccessResponse(
+		samlSession.Entry.ServiceProviderID,
+		authenticatedUserID,
+		authnRequest,
+	)
+	if err != nil {
+		panic(err)
```

Comment:
And do not panic here.

Reply:
It is an unexpected error here so I panic here.